### PR TITLE
[registration] Removing deprecated method `setInputCloud`  from public API

### LINKED ISSUE #2724: Removing deprecated methods changed behavior
## Your Environment

* Operating System and version: OSX
* Compiler: Clang
* PCL Version: 1.9

## Context

Commit 16d3a26d9fbd890beb87bbeab8b0d39936ec2837 broke existing behavior of 1.8.  IterativeClosestPoint::setInputCloud() was removed.  It had been marked as deprecated.  But since setInputCloud() exists in PCLBase, a base class of IterativeClosestPoint, code continued to compile and run with no warnings after IterativeClosestPoint::setInputCloud() was removed.  This is bad.

## Expected Behavior

Removing a deprecated function should cause code using the deprecated function not to build.

## Code to Reproduce

The code still builds without warning -- it just runs improperly.

## Possible Solution

Replace setInputCloud() and have it throw an exception.  Or it could be renamed.  Someone who knows more about the code base would have to decide such things.

### PR DESCRIPTION
**Pull request by kunaltyagi**
Fixes #2724 by making compiler emit compile error if previously deprecated method is used

### CODE REVIEW

**Comment by kunaltyagi** [comment]
I'm not sure that this doesn't break behavior of downstream users in subtle ways.

File: registration/include/pcl/registration/registration.h [hunk 1]
```diff
@@ -609,6 +609,12 @@ namespace pcl
     private:
       /** \brief The point representation used (internal). */
       PointRepresentationConstPtr point_representation_;
+
+      /**
+       * \brief Remove from public API in favor of \ref setInputSource
+       */
+      using PCLBase<PointSource>::setInputCloud;
```
Comment (SergioRAgostinho) [inline registration/include/pcl/registration/registration.h:616]: Semantically, [deleting the method](https://en.cppreference.com/w/cpp/language/function#Deleted_functions) would better express the right intent here. Move it to the original scope and delete it.

**Comment by SergioRAgostinho** [comment]
Merge at your convenience if the CI is green.

File: registration/include/pcl/registration/registration.h (hunk 1 above)
Comment (kunaltyagi) [reply registration/include/pcl/registration/registration.h]: https://godbolt.org/z/qYpD-u

Can't delete a virtual function, apparently. Any other syntax that can help achieve this?

**Comment by taketwo** [comment]
According to [this](https://stackoverflow.com/a/24610165/1525865), deleting a virtual function in deriving classes is not allowed by the standard. Changing visibility (as originally proposed by @kunaltyagi) is one of the listed workarounds. Another one is to add a `static_assert` to the override. Perhaps it would be a more user-friendly approach because we can explain the user what to do in the assertion message.

**Comment by kunaltyagi** [comment]
I tested out a snippet in [Godbolt](https://godbolt.org/z/45QNAP) and it didn't throw an error. Please show me the error in my ways 😆. A valid snippet would be very welcome indeed.

**Comment by SergioRAgostinho** [comment]
Your example was broken. You cannot have a templated virtual function, so the function you declared was not overriding the original one. See the modified snippet https://godbolt.org/z/LdkanP

**Comment by kunaltyagi** [comment]
`static_assert(false,...)` will fail immediately. The template is there to delay it from being evaluated by the compiler

See [snippet](https://godbolt.org/z/7ZUeMQ)

**Comment by SergioRAgostinho** [comment]
I took another look at the SO thread and the associated comments and they flag that the method can still be called from the base class, with the static_assert option. This happens because we're redefining the method in the derived class, not really overriding it.

Both solutions on the table (static assert and moving th private) suffer from the same, there's no way to prevent the call from a base pointer, because the base pointer has the method implemented.

Given the circumstances I vote to move it to private but keep the original implementation we deleted. There's no point in having a static deprecated warning because we'll never be able to delete the method properly, so we should just throw that warning at runtime.

**Comment by kunaltyagi** [comment]
PTAL. Updated to use private + runtime warning.

**Pull request merged**